src/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -1422,17 +1428,19 @@
   "Reactivate the given host, only if it was deactivated or expired before
   `time`.  Returns true if the node is activated, or if it was already active.
 
-  Adds the host to the database if it was not already present."
+   Adds the host to the database if it was not already present."
   [certname :- String
    time :- pls/Timestamp]
-  (when-not (certname-exists? certname)
-    (add-certname! certname))
-  (let [timestamp (to-timestamp time)
-        replaced  (jdbc/update! :certnames
-                                {:deactivated nil, :expired nil}
-                                ["certname=? AND (deactivated<? OR expired<?)"
-                                 certname timestamp timestamp])]
-    (pos? (first replaced))))
+  (let [{:keys [deactivated expired] :as metadata} (first (certname-metadata certname))]
+    (when-not metadata
+      (add-certname! certname))
+    (let [timestamp (to-timestamp time)
+          replaced  (when (or deactivated expired)
+                      (jdbc/update! :certnames
```

Comment:
That seems like the more safe thing to do, especially since you're making that RR change anyway.